src/core/infer.ml

```diff
@@ -1,216 +1,201 @@
-      (* [get_prod f typ] returns the domain and codomain of [t] if [t] is a
-         product. If [t] is a metavariable, then it instantiates it with a
-         product and calls [get_prod f typ] again. Otherwise, it calls [f
-         typ]. *)
-      let rec get_prod f typ =
-        if !log_enabled then log_infr "get_prod %a" pp_term typ;
-        match unfold typ with
-        | Prod(a,b) -> (a,b)
-        | Meta(m,_) -> set_to_prod p m; get_prod f typ
-        | _ -> f typ
+      let a = try Ctxt.type_of x ctx with Not_found -> assert false in
+      (t, a)
+  | Symb s -> (t, !(s.sym_type))
+  | Plac (true, name) ->
+      let m = LibMeta.make pb ?name ctx mk_Type in
+      (m, mk_Type)
+  | Plac (false, name) ->
+      let mt = LibMeta.make pb ctx mk_Type in
+      let m = LibMeta.make pb ?name ctx mt in
+      (m, mt)
+  (* All metavariables inserted are typed. *)
+  | (Meta (m, ts)) as t ->
+      let rec ref_esubst i range =
+        (* Refine terms of the explicit substitution. *)
+        if i >= Array.length ts then range else
+          match unfold range with
+          | Prod(ai, b) ->
+              ts.(i) <- force pb ctx ts.(i) ai;
+              let b = Bindlib.subst b ts.(i) in
+              ref_esubst (i + 1) b
+          | _ ->
+              (* Meta type must be a product of arity greater or equal
+                 to the environment *)
+              assert false
+      in
+      let range = ref_esubst 0 !(m.meta_type) in
+      (t, range)
+  | LLet (t_ty, t, u) ->
+      (* Check that [a] is a type, and refine it. *)
+      let t_ty, _ = type_enforce pb ctx t_ty in
+      (* Check that [t] is of type [a'], and refine it *)
+      let t = force pb ctx t t_ty in
+      (* Unbind [u] and get new context with [x: t_ty ≔ t] *)
+      let x, u, ctx = Ctxt.unbind ctx t_ty (Some t) u in
+      (* Infer type of [u'] and refine it. *)
+      let u, u_ty = infer pb ctx u in
+      let u_ty = Bindlib.(u_ty |> lift |> bind_var x |> unbox) in
+      let u_ty = Bindlib.subst u_ty t in
+      let u = Bindlib.(u |> lift |> bind_var x |> unbox) in
+      (mk_LLet (t_ty, t, u), u_ty)
+  | Abst (dom, b) ->
+      (* Domain must by of type Type, we don’t use [type_enforce] *)
+      let dom = force pb ctx dom mk_Type in
+      let x, b, ctx = Ctxt.unbind ctx dom None b in
+      let b, range = infer pb ctx b in
+      let b = Bindlib.(lift b |> bind_var x |> unbox) in
+      let range = Bindlib.(lift range |> bind_var x |> unbox) in
+      (mk_Abst (dom, b), mk_Prod (dom, range))
+  | Prod (dom, b) ->
+      (* Domain must by of type Type, we don’t use [type_enforce] *)
+      let dom = force pb ctx dom mk_Type in
+      let x, b, ctx = Ctxt.unbind ctx dom None b in
+      let b, b_s = type_enforce pb ctx b in
+      let s =
+        match unfold b_s with
+        | Type -> mk_Type
+        | Kind -> mk_Kind
+        | b_s ->
+            Error.wrn None
+              "Type error, sort mismatch: there is no rule of the form \
```

Comment:
precise: rule -> typing rule
But why do you speak of typing rules for products here? Isn't it just the case that the type of b is not a sort?

Reply:
The type of `b` must always be a sort, doesn't it?

Reply:
Right.